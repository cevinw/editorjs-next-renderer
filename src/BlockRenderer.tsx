import React from 'react';
import AlertBlock from "./BlockParsers/alert";
import ParagraphBlock from "./BlockParsers/paragraph";
import CodeBlock, {CodeLanguage} from "./BlockParsers/code";
import DelimiterBlock from "./BlockParsers/delimiter";
import HeaderBlock from "./BlockParsers/header";
import ImageBlock from "./BlockParsers/image";
import ColumnsBlock from "./BlockParsers/columns";
import QuoteBlock from "./BlockParsers/quote";
import TableBlock from "./BlockParsers/table";
import ListBlock from "./BlockParsers/list";
import ErrorBlock from "./BlockParsers/error";


export interface ArticleRenderProps {
    data: OutputData;
    config?: BlockRendererConfig
}

export interface BlockRendererConfig {
    alert?: { className: string },
    code?: {
        className?: string,
        codeStyle?: { [key: string]: React.CSSProperties }
        languages?: CodeLanguage[],
        showLineNumbers?: boolean,
    },
    table?: { className: string },
    image?: { className: string },
    header?: { className: string },
    paragraph?: { className: string },
    list?: { className: string },
    quote?: { className: string },
    delimiter?: { className: string }
}

export interface OutputBlockData<Data extends object = any> {

    id?: BlockId;
    type: string;
    data: BlockToolData<Data>;
    tunes?: { [name: string]: BlockTuneData };
}

export interface OutputData {

    version?: string;
    time?: number;
    blocks: OutputBlockData[];
}

export type BlockId = string;
export type BlockToolData<T extends object = any> = T;
export type BlockTuneData = any;

const BlockRenderer = ({data, config}: ArticleRenderProps) =>
    data.blocks.map((blockData: OutputBlockData) => parseData(blockData, config));

const parseData = (item: OutputBlockData, config?: BlockRendererConfig): React.JSX.Element => {
    return {
        "code": <CodeBlock key={item.id} item={item} {...config?.code}/>,
        "paragraph": <ParagraphBlock key={item.id} item={item} {...config?.paragraph}/>,
        "header": <HeaderBlock key={item.id} item={item} {...config?.header}/>,
        "alert": <AlertBlock key={item.id} item={item} {...config?.alert}/>,
        "columns": <ColumnsBlock key={item.id} item={item} config={config}/>,
        "quote": <QuoteBlock key={item.id} item={item} {...config?.quote}/>,
        "table": <TableBlock key={item.id} item={item} {...config?.table}/>,
        "list": <ListBlock key={item.id} item={item} {...config?.list}/>,
        "delimiter": <DelimiterBlock key={item.id} item={item} {...config?.delimiter}/>,
        "image": <ImageBlock key={item.id} item={item} {...config?.image}/>
    }[item.type] || <ErrorBlock item={item} key={item.id}/>
}

export default BlockRenderer;