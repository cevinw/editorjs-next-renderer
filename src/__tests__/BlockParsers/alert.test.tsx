import {describe, expect, test} from "vitest";
import {render} from "@testing-library/react";
import React from "react";
import AlertBlock, {EditorJsAlert} from "../../BlockParsers/alert";
import {OutputBlockData} from "../../BlockRenderer";

describe("Alert Block tests", () => {
    test("Render alert snapshot test", async () => {

        const alertBlockData : OutputBlockData<EditorJsAlert> = {
            data: {
                type: "success",
                align: "center",
                message: "test"
            },
            type: "alert",
            id: "id",

        }

        const {container} = render(<AlertBlock item={alertBlockData}/>);
        expect(container).toMatchSnapshot();
    })

    test.each([
        ["success", "bg-green-100"],
        ["danger", "bg-red-200"],
        ["warning", "bg-orange-100"],
        ["light", "bg-white"],
        ["dark", "bg-gray-900"],
        ["primary", "bg-gray-300"],
        ["info", "bg-gray-300"]

    ])("Render alert type test", async (type: string, expected: string) => {
        const alertBlockData : OutputBlockData<EditorJsAlert> = {
            data: {
                type: type,
                align: "center",
                message: "test"
            },
            type: "alert",
            id: "id",

        }
        const {container} = render(<AlertBlock item={alertBlockData}/>);
        expect(container.querySelector("section")?.className).contains(expected);

    })
})
