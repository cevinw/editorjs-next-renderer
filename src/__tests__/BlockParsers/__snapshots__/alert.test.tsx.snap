// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert Block tests > Render alert snapshot test 1`] = `
<div>
  <figure
    class="p-2 mt-2 px-4 flex bg-opacity-50 shadow-sm rounded-lg bg-green-100 text-green-800"
  >
    <figcaption
      class="text-center"
    >
      test
    </figcaption>
  </figure>
</div>
`;
